numpy/lib/tests/test_histograms.py

```diff
@@ -299,6 +299,30 @@ def test_datetime(self):
         assert_equal(d_edge.dtype, dates.dtype)
         assert_equal(t_edge.dtype, td)
 
+    def test_precision_lower_bound(self):
+        tiny_shift = 1e-8
+        arr = np.array([1.0], np.float32)
+        range = np.array([1.0 + tiny_shift, 2.0], np.float64)
+
+        # previously crashed
+        count, x_loc = np.histogram(arr, bins=1, range=range)
+
+        # gh-10322 means that the float64 decays to float32.
+        assert_equal(x_loc.dtype, np.float32)
+        assert_equal(count, [1])
+
+    def test_precision_upper_bound(self):
+        tiny_shift = 1e-8
+        arr = np.array([1.0], np.float32)
+        range = np.array([0.0, 1.0 - tiny_shift], np.float64)
+
+        # previously crashed
+        count, x_loc = np.histogram(arr, bins=1, range=range)
+
+        # gh-10322 means that the float64 decays to float32.
+        assert_equal(x_loc.dtype, np.float32)
+        assert_equal(count, [1])
```

Comment:
In #9189, the results here were `np.float64` and `[0]`, which are self-consistent, but at odds with #10322.

Reply:
I am a bit at odds with this code... I have modified `np.result_type` to *only* special case Python float, integer, (and complex).  Even adding a warning on it about the switch, there are only a handful of places in NumPy such a change. And all of them seem to explicitly test corner cases (except possibly the default value in `np.select` which arguably should be modified to `None` which then could use `result.dtype.type()` as default).

So if you look at the code chunk with comments below, the `result_type` returns the "imprecise" version currently. If that would opt into a more precise version, this changes to `[0]` as commented.  However, warnings about such a change seems pretty drastic/weird and potentially impossible to get right?: We have to pull the plug on gh-10322 eventually!

 In longer words: I am in a conundrum, that I we need to fix a behaviour here (potentially with semi-public functions to opt-in to the"future" behaviour early).

Reply:
Maybe its not actually a big issue, we just a lot of warnings and users can only avoid them by casting their inputs. The problem may just be that I pulled the plug on `np.result_type` for testing, but did not modify the ufunc code yet (and that is too big of a throw to do at once).

Then, in the future, this *will* switch to `0` when bins are typed at least.